File: VIDEO/app/utils/alert_class_filter.py
"""算法任务告警触发类别过滤。"""
import json
from typing import Any, List, Optional


def normalize_class_name(class_name: str) -> str:
    return str(class_name or '').strip().lower().replace('-', '_').replace(' ', '_')


def parse_alert_class_names(raw: Any) -> List[str]:
    if raw is None:
        return []
    if isinstance(raw, list):
        names = raw
    elif isinstance(raw, str):
        text = raw.strip()
        if not text:
            return []
        try:
            parsed = json.loads(text)
            names = parsed if isinstance(parsed, list) else []
        except (TypeError, ValueError, json.JSONDecodeError):
            names = [text]
    else:
        return []
    result = []
    seen = set()
    for name in names:
        label = str(name or '').strip()
        if not label:
            continue
        key = normalize_class_name(label)
        if key in seen:
            continue
        seen.add(key)
        result.append(label)
    return result


def filter_detections_for_alert(detections: list, alert_class_names: Any) -> list:
    """按任务配置的告警标签过滤检测结果；未配置标签时保持兼容（任意检测均可告警）。"""
    if not detections:
        return []
    allowed = parse_alert_class_names(alert_class_names)
    if not allowed:
        return list(detections)
    allowed_set = {normalize_class_name(name) for name in allowed}
    return [
        det for det in detections
        if normalize_class_name(det.get('class_name', '')) in allowed_set
    ]


def get_task_alert_class_names(task_config: Any) -> Optional[Any]:
    if task_config is None:
        return None
    return getattr(task_config, 'alert_class_names', None)

File: DEVICE/iot-sink/iot-sink-biz/src/main/java/com/basiclab/iot/sink/util/AlertClassFilter.java
package com.basiclab.iot.sink.util;

import org.springframework.util.StringUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

/**
 * 算法任务告警触发类别过滤（与 VIDEO alert_class_filter.py 规则一致）。
 */
public final class AlertClassFilter {

    private AlertClassFilter() {
    }

    public static String normalizeClassName(String className) {
        if (!StringUtils.hasText(className)) {
            return "";
        }
        return className.trim().toLowerCase(Locale.ROOT)
                .replace('-', '_')
                .replace(' ', '_');
    }

    @SuppressWarnings("unchecked")
    public static List<String> parseAlertClassNames(Object raw) {
        if (raw == null) {
            return Collections.emptyList();
        }
        List<Object> names;
        if (raw instanceof List) {
            names = (List<Object>) raw;
        } else if (raw instanceof String) {
            String text = ((String) raw).trim();
            if (!StringUtils.hasText(text)) {
                return Collections.emptyList();
            }
            if (text.startsWith("[")) {
                try {
                    List<Object> parsed = com.basiclab.iot.common.utils.json.JsonUtils.parseObject(
                            text, new com.fasterxml.jackson.core.type.TypeReference<List<Object>>() {});
                    names = parsed != null ? parsed : Collections.singletonList(text);
                } catch (Exception e) {
                    names = Collections.singletonList(text);
                }
            } else {
                names = Collections.singletonList(text);
            }
        } else {
            return Collections.emptyList();
        }

        Set<String> seen = new LinkedHashSet<>();
        List<String> result = new ArrayList<>();
        for (Object item : names) {
            if (item == null) {
                continue;
            }
            String label = String.valueOf(item).trim();
            if (!StringUtils.hasText(label)) {
                continue;
            }
            String key = normalizeClassName(label);
            if (seen.contains(key)) {
                continue;
            }
            seen.add(key);
            result.add(label);
        }
        return result;
    }

    public static List<Map<String, Object>> filterDetectionsForAlert(
            List<Map<String, Object>> detections,
            Object alertClassNames) {
        if (detections == null || detections.isEmpty()) {
            return Collections.emptyList();
        }
        List<String> allowed = parseAlertClassNames(alertClassNames);
        if (allowed.isEmpty()) {
            return detections;
        }
        Set<String> allowedSet = new HashSet<>();
        for (String name : allowed) {
            allowedSet.add(normalizeClassName(name));
        }
        List<Map<String, Object>> filtered = new ArrayList<>();
        for (Map<String, Object> det : detections) {
            if (det == null) {
                continue;
            }
            Object rawName = det.get("class_name");
            if (rawName == null) {
                rawName = det.get("className");
            }
            String className = rawName != null ? String.valueOf(rawName) : "unknown";
            if (allowedSet.contains(normalizeClassName(className))) {
                filtered.add(det);
            }
        }
        return filtered;
    }
}
